[process-agent] Better process collection on Windows

xornivore opened the pull request:
### What does this PR do?

- Adds probe implementation based on using Windows PDH performance counter APIs which has has the following benefits over the existing `toolhelp` API collector:
  - No need to use process handles to collect process stats
  - Largely reduced number of API calls - we can collect data for all processes using very few calls to get PDH formatted value arrays.
  - New implementation can be turned on by setting `process_config.windows.use_perf_counters`, however the previous check implementation based on the `toolhelp` API will remain the default in the next release.

- Redefines `procutil.Probe` as an interface to allow for multiple implementations for the same OS.
- Consolidate `Process` and `RTProcess` checks:
    - Defines a new interface for a check that can run both regular and RT data collection (`CheckWithRealTime`)
    - Implements a custom check runner that performs skips of RT interval when RT is not enabled.
    - Consolidates functionality in `Process` and `RTProcess` allowing to use the same instance of `procutil.Probe`.
    - Eliminates the need for synchronization between regular and RT process data collection.
  
### Motivation

- Better, more maintainable process collection on Windows.
- Eliminate the need for `procutil.Probe` implementations to be thread safe and eliminate the synchronization between `RTProcess` and `Process` checks. 

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

These are pretty large changes that touch process collection code for multiple platforms.

- Validate that process and RT collection works correctly in all supported platforms.
- Validate the use of perf counter based probe and compare it to the existing legacy toolhelp implementation across supported versions of Windows.
- Validate that the `Connections` check can still properly use create time info from the `Process` check.

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

xornivore committed:
[process-agent] WIP improved process collection on Windows using PDH

- Adds probe implementation based on using Windows PDH performance counter APIs
- Redefine Probe as interface to allow for multiple implementations for the same OS.

Commit:
Consolidate Process and RTProcess checks

- Define a new interface for a check that can run both regular and RT data collection (`CheckWithRealTime`)
- Implement a custom check runner that performs skips of RT interval when RT is not enabled.
- Consolidate functionality in Process and RTProcess allowing to use the same instance of procutil.Probe.
- Eliminate the need for synchronization between regular and RT process collection.

mx-psi commented on tasks/go.py:
```diff
@@ -21,14 +21,15 @@
 
 # List of modules to ignore when running lint
-MODULE_WHITELIST = [
+MODULE_ALLOWLIST = [
```
Can you do this name change (and the change in the printed message) in a separate PR?

xornivore replied:
Sounds good, I will extract this into a separate PR.

mx-psi commented on tasks/go.py:
```diff
@@ -21,14 +21,15 @@
     "pdh.go",
     "pdh_amd64.go",
     "pdh_386.go",
+    "pdhformatter.go",
```
Why is this needed?

xornivore replied:
So I was able to work around the issue causing the warning about `misuse of unsafe.Pointer`. At first I misinterpreted as belonging to the lint output (hence the changes to this invoke task), but in fact it is from `go vet` and there isn't a good way to mark exclusions for it, `// go vet: .. ` does seem to not work. Afaict the only option as of today is to build with `!dovet`. To summarize - once I confirm I should be able to remove addition of this file to the allow list.

mx-psi replied:
Yeah, `go vet` doesn't have an exclusion mechanism sadly :( we can use the skip list if you don't find another way around it

just-chillin commented on pkg/process/procutil/process_windows.go:
```diff
@@ -0,0 +1,514 @@
+	// 0x1000 is PROCESS_QUERY_LIMITED_INFORMATION, but that constant isn't
+	//        defined in x/sys/windows
+	// 0x10   is PROCESS_VM_READ
+	procHandle, err := windows.OpenProcess(0x1010, false, uint32(pid))
```
nit: Can we create a constant for 0x1010 and 0x10?

xornivore replied:
yep, this is duplicated in a couple of places, probably best done in a separate PR, it belongs in the `pkg/util/winutil`.

just-chillin commented on pkg/util/winutil/pdhutil/pdh.go:
```diff
@@ -132,6 +132,42 @@ const (
+	CounterAllProcessWorkingSetPrivate  = `\Process(*)\Working Set - Private`
+)
+
+type PDH_FMT_COUNTERVALUE_ITEM struct {
```
Can we add a comment explaining what this does, and linking it to the microsoft docs for the one in `pdh.h`?

https://docs.microsoft.com/en-us/windows/win32/api/pdh/ns-pdh-pdh_fmt_countervalue_item_a

xornivore replied:
I think generally this is clear we are dealing with PDH APIs, but adding a comment to the struct makes sense.

just-chillin commented on pkg/process/procutil/process_linux.go:
```diff
@@ -53,34 +53,37 @@ type statInfo struct {
-// Probe is a service that fetches process related info on current host
-type Probe struct {
+// probe is a service that fetches process related info on current host
+type probe struct {
```
I think we should probably rename this to something more meaningful.

xornivore replied:
For the sake of keeping this PR smaller probably worth splitting that, though I think pondering a good alternative might take time - collector (taken), enumerator (odd), if you have a suggestion feel free to post here.

just-chillin commented on pkg/util/winutil/pdhutil/pdhformatter.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	var items []PDH_FMT_COUNTERVALUE_ITEM
+	hdrItems := (*reflect.SliceHeader)(unsafe.Pointer(&items))
```
The inner rust developer in me sets off alarms whenever I see an unsafe pointer dereference. Is it possible for any part of this to fail?

I would add a comment explaining why this is safe to do, and how we know that we won't segfault whenever you use unsafe.

xornivore replied:
Adding a comment sounds like a good idea.

xornivore commented on pkg/process/checks/allprocesses_windows.go:
```diff
@@ -28,6 +29,7 @@ var (
 
 	// XXX: Cross-check state is stored globally so the checks are not thread-safe.
 	cachedProcesses = map[uint32]*cachedProcess{}
+	// TODO: remove locking here
```
With these changes we no longer have to guard cachedProcesses with a lock because we are using the same goroutine to run both regular and RT checks (there is no contention).

xornivore commented on pkg/process/checks/process.go:
```diff
@@ -33,9 +34,7 @@ type ctrProcMsgFactory func([]*model.Process, ...*model.Container) *model.Collec
 // for live and running processes. The instance will store some state between
 // checks that will be used for rates, cpu calculations, etc.
 type ProcessCheck struct {
```
I suspect rethinking that singleton pattern is a separate effort - they are in the "all checks" slice, and many other checks would need to change.

xornivore commented on pkg/process/checks/process.go:
```diff
@@ -22,7 +21,9 @@ import (
 
 // Process is a singleton ProcessCheck.
-var Process = &ProcessCheck{probe: procutil.NewProcessProbe()}
+var Process = &ProcessCheck{}
```
Agreed, in principle, in fact each check already implements the exported `Check` interface and these are currently consumed in [cmd/process-agent](https://github.com/DataDog/datadog-agent/blob/main/cmd/process-agent/collector.go#L74) via the [All](https://github.com/DataDog/datadog-agent/blob/main/pkg/process/checks/checks.go#L20) slice of `Check` checks. If we decide to change the fact that the names are exported, it's probably a change better consolidated to a separate PR because we'd want to rename other checks as well.

xornivore commented on pkg/process/checks/process.go:
```diff
@@ -45,16 +44,26 @@ type ProcessCheck struct {
 
 // Init initializes the singleton ProcessCheck.
-func (p *ProcessCheck) Init(_ *config.AgentConfig, info *model.SystemInfo) {
+func (p *ProcessCheck) Init(cfg *config.AgentConfig, info *model.SystemInfo) {
```
Do you mean an equivalent of `NewProcessCheck` where we create and initialize it?

Commit:
Address review comments in pdhutil

shang-wang commented on cmd/process-agent/main_common.go:
```diff
@@ -239,40 +240,79 @@ func debugCheckResults(cfg *config.AgentConfig, check string) error {
+	}
+
+	for _, r := range runs {
+		if r.CheckName == ch.RealTimeName() {
```
nit: this can be removed.

shang-wang commented on pkg/process/checks/process.go:
```diff
@@ -110,14 +131,13 @@ func (p *ProcessCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.Mess
 
 	// stores lastPIDs to be used by RTProcess
-	lastPIDs := make([]int32, 0, len(procs))
+	p.lastPIDs = make([]int32, 0, len(procs))
```
nit: since everything is synchronous, we could use `p.lastPIDs = p.lastPIDs[:0]` to avoid allocating new memory.

shang-wang commented on pkg/process/checks/process.go:
```diff
@@ -78,9 +90,17 @@ func (p *ProcessCheck) RealTime() bool { return false }
+	if len(results) == 0 {
+		return nil, nil
+	}
+	return results[0].Messages, nil
```
Sorry I was a bit confused by this, sounds like the `results` is only going to have either one or two items depend on whether it's RT or not? If so why not creating a struct? I have not reviewed the whole thing so I could be missing something.

xornivore replied:
I agree with you, it's quite clunky as is in the end, I will change results to be a struct with two fields.

shang-wang commented on pkg/process/checks/runner.go:
```diff
@@ -0,0 +1,99 @@
+
+type runnerWithRealTime struct {
+	RunnerConfig
+	ratio      int
+	counter    int
```
Can we have some comment on how they are used?

shang-wang commented on pkg/process/checks/runner.go:
```diff
@@ -0,0 +1,99 @@
+	RunnerConfig
+	ratio      int
+	counter    int
+	newTicker  func(d time.Duration) *time.Ticker
+	stopTicker func(t *time.Ticker)
```
nit: why can't the `newTicker` be a `ticker` property of `runnerWithRealTime`? And why `stopTicker` has to be a property instead?

xornivore replied:
Practically it's just dependency injection for unit tests, `ticker` is re-created on RT change in the goroutine, and stopping a ticker panics if it's not initialized (in cases when we inject it, it's not properly initialized, just the channel is initialized.

shang-wang commented on pkg/process/config/yaml_config.go:
```diff
@@ -187,6 +187,11 @@ func (a *AgentConfig) LoadProcessYamlConfig(path string) error {
 	}
 
+	// Windows: Controls using the new check based on performance counters PDH APIs
+	if usePerfCountersKey := key(ns, "windows", "use_perf_counters"); config.Datadog.IsSet(usePerfCountersKey) {
```
nit: the config has to be added as `SetKnown()` in `pkg/config/config.go` otherwise it'll report warning when agent starts and parsing this config.

xornivore replied:
That's a good point, I forgot about it.

Commit:
Address review feedback

Commit:
Fix nil dereferences.

Commit:
Fix bugs in performance counter to stat conversions

- Type cast entire item on enumeration in the formatter
- Map `float64` and `uint64` using additional helpers
- Add traces for values and use declared funcs instead of anonymous.
- Ignore `Idle` and `_Total` counter instances.

Commit:
Address more TODOs, cleanup legacy win probe

- Clean up synchronization in the legacy win probe
- Address comment TODO in realtime process check
- Ensure groupID is unique between invocations of standard and realtime checks.
- Handle error on counter values

Commit:
Fix bug with RSS/VMS assignment in the new check

Commit:
Fix calculation of CPU percent for the new check

- PDH counters only report percent and not times.
- The times which we currently report (not percent) are actually not utilized on the backend so this is safe.
- Adding both ways to handle the percent calculation - either taking as is or existing (based on CPU time increase divided by delta).

xornivore commented on pkg/process/checks/allprocesses_windows.go:
```diff
@@ -25,32 +25,8 @@ var (
-	haveWarnedNoArgs    = false
 )
 
-type SystemProcessInformation struct {
```
Was not used here

ISauve reviewed (approved):
very shallow review as I only looked at the file owned by the `network-agent` team - but LGTM!

djmitche reviewed (approved):
Also a shallow review of the winutil bits. The PR description sounds good, especially the use of a feature gate.

wiyu commented on pkg/process/checks/process.go:
```diff
@@ -133,7 +151,15 @@ func (p *ProcessCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.Mess
+			p.realtimeLastCPUTime = p.lastCPUTime
+			p.realtimeLastProcs = procsToStats(p.lastProcs)
+			p.realtimeLastCtrRates = p.lastCtrRates
+			p.realtimeLastRun = p.lastRun
```
nit: Would be nice if the storing of the last checks was in a function, but maybe in a future PR.

xornivore replied:
Agreed, I will do some refactoring in a separate PR along with more tests, now that the probe is a mockable interface it should be fairly easy to do that.

shang-wang replied:
Even better we could wrap the `last***` and `realtimeLast***` variables into structs.

xornivore replied:
yep, that's a great idea and it would prevent obvious typos in some cases

wiyu commented on pkg/process/checks/process.go:
```diff
@@ -316,17 +393,29 @@ func fmtProcesses(
 		fp.Cmdline = cfg.Scrubber.ScrubProcessCommand(fp)
 
+		var ioStat *model.IOStat
+		if fp.Stats.IORateStat != nil {
```
What's the difference btw this `IORateStat` and the `IOStat`?

xornivore replied:
I had to abstract the rate (in /sec measurements) vs increasing values (that we convert to rates):
```
// IOCountersStat holds IO metrics for a process
type IOCountersStat struct {
	ReadCount  int64
	WriteCount int64
	ReadBytes  int64
	WriteBytes int64
}

// IOCountersRateStat holds IO metrics for a process represented as rates (/sec)
type IOCountersRateStat struct {
	ReadRate       float64
	WriteRate      float64
	ReadBytesRate  float64
	WriteBytesRate float64
}
```

shang-wang replied:
Took a minute for me to realized that `IOCountersRateStat` is used in Windows and the old `IOCountersStat` is used in others(am I right?). If it's the case I feel like explicitly calling them out as comments on the model file would make it easier to understand.

Commit:
Use image path as fallback when unable to open process

- In certain cases we are unable to open the process to read memory (PEB contains command args)
- Use image path as a fallback

Commit:
Small performance improvements and cleanup

- Get rid of tracing for now - will add PID filter in a future PR to be able to trace.
- Change log to Debug level for counter error (can be expected).
- Deduplicate code between process/checks and process/procutil.

Commit:
Add handling of misconfigured processs check intervals

mbotarro commented on cmd/process-agent/collector.go:
```diff
@@ -105,9 +105,43 @@ func (l *Collector) runCheck(c checks.Check, results *api.WeightedQueue) {
+			return
+		}
+		l.messagesToResults(start, withRealTime.Name(), run.Standard, results)
+		l.messagesToResults(start, withRealTime.RealTimeName(), run.RealTime, results)
```
Is the same `api.WeightedQueue` passed to both regular and RT checks? https://github.com/DataDog/datadog-agent/pull/8875 created a separate queue for RT checks. I might be missing something but it seems that we're using the same queue for both checks now.

xornivore replied:
Nice catch, yep, this PR predates #8875, I will fix this.

xornivore replied:
done.

mbotarro commented on cmd/process-agent/collector.go:
```diff
@@ -105,9 +105,43 @@ func (l *Collector) runCheck(c checks.Check, results *api.WeightedQueue) {
+		d := time.Since(start)
+		switch {
+		case runCounter < 5:
+			log.Infof("Finished %s check #%d in %s", c.Name(), runCounter, d)
```
should we only increment `runCounter` when the normal check is run?

xornivore replied:
We do check for it [here](https://github.com/DataDog/datadog-agent/pull/8846/files/08c387d53948737801e468cd267263f3b15968e2#diff-993d2491e7efac8e13349ef7859c375a68e328aa5415289b699a6da0cc2637c3R126)

mbotarro replied:
If `runCheck` is called for both standard and RT checks, won't `runCounter` be increased by RT checks? 
https://github.com/DataDog/datadog-agent/blob/08c387d53948737801e468cd267263f3b15968e2/cmd/process-agent/collector.go#L99-L103

xornivore replied:
yes, but RT check in that case is an illusion because we are just copying stats from the standard check

Commit:
Use RT queue for CheckWithRealtime realtime results

just-chillin commented on pkg/process/procutil/process_model.go:
```diff
@@ -33,13 +62,55 @@ type Stats struct {
+	}
+	if s.CtxSwitches != nil {
+		copy.CtxSwitches = &NumCtxSwitchesStat{}
+		*copy.CtxSwitches = *s.CtxSwitches
```
Why are we writing to the struct twice?

xornivore replied:
we are initializing the pointer, then copying the data to the struct.

just-chillin reviewed (approved):
LGTM

Commit:
Fix panic on fetch pids process check => connections

- Connections check expects the map to be non-nil and updates with 0 values if pid is not found.

mbotarro commented on cmd/process-agent/main_common.go:
```diff
@@ -320,40 +321,82 @@ func debugCheckResults(cfg *config.AgentConfig, check string) error {
+		}
+	)
+
+	if _, err := ch.RunWithOptions(cfg, nextGroupID, options); err != nil {
+		return fmt.Errorf("collection error: %s", err)
+	}
```
nit: should we add a comment to explain why we need to run the check twice?

xornivore replied:
good point I will add that 🙇‍♂️

shang-wang commented on pkg/process/procutil/process_model.go:
```diff
@@ -94,6 +171,14 @@ type IOCountersStat struct {
 	WriteBytes int64
 }
 
+// IOCountersRateStat holds IO metrics for a process represented as rates (/sec)
```
nit: let's not insert this definition between `IOCountersStats` and `func (i *IOCountersStat) IsZeroValue()` which ideally should be together.

xornivore replied:
Makes sense, it's an oversight on my part.

shang-wang commented on pkg/process/checks/runner.go:
```diff
@@ -0,0 +1,99 @@
+	if checkInterval <= rtInterval {
+		return -1, errors.New("check interval should be larger than RT interval")
+	}
+	if checkInterval%rtInterval != 0 {
```
Just want to make sure that this is defensive programming right? Because I was checking the config section, and the logic would use default intervals if the config doesn't make sense.

xornivore replied:
yes, it is defensive coding

shang-wang commented on pkg/process/config/yaml_config.go:
```diff
@@ -94,6 +94,19 @@ func (a *AgentConfig) LoadProcessYamlConfig(path string) error {
 	a.setCheckInterval(ns, "process_realtime", RTProcessCheckName)
 	a.setCheckInterval(ns, "connections", ConnectionsCheckName)
 
+	if a.CheckIntervals[ProcessCheckName]%a.CheckIntervals[RTProcessCheckName] != 0 {
```
Should there be code to also make sure regular check interval is always greater than RT check?

xornivore replied:
Great point I will add that.

Commit:
Address review feedback and rename default probe

- Add comment explaining why the check needs to be run twice.
- Reorganize declaration of IOCountersRateStat.
- Ensure that Process check interval is greater than RTProcess check interval.

shang-wang commented on cmd/process-agent/collector.go:
```diff
@@ -315,6 +329,71 @@ func (l *Collector) run(exit chan struct{}) error {
+	return l.basicRunner(c, results, exit), nil
+}
+
+func (l *Collector) basicRunner(c checks.Check, results *api.WeightedQueue, exit chan struct{}) func() {
```
`legacyRunner`?

xornivore replied:
It's not really legacy, maybe `simpleRunner`? because there are checks that will always be using this (e.g. there's no realtime`pod` or `connections`).

shang-wang replied:
Ah you are right, I missed the non-process checks.

xornivore replied:
No problem, it would be nice to move some of this collector code under `pkg/process` including the runner code here. It's a bit busy.

shang-wang commented on pkg/process/procutil/process_windows.go:
```diff
@@ -0,0 +1,609 @@
+
+// GetUsernameForProcess returns username for a process
+func GetUsernameForProcess(h windows.Handle) (name string, err error) {
+	name = ""
```
nit: these are not needed.

shang-wang reviewed (approved):
🔥

Commit:
Remove unnecessary empty string initialization

Commit:
Fix probe type (result of merge issue)

Commit:
Add const for discovery check

Commit:
Correct interface/implementation signature

Commit:
Correctly handle collectStats in new win process probe
